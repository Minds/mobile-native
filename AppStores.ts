import { createLegacyStores, TLegacyStores } from './src/common/contexts';

let stores: any;

export const getStores = function (): TLegacyStores {
  if (!stores) {
    stores = createLegacyStores();
  }
  return stores;
};
