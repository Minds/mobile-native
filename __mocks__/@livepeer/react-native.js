// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
