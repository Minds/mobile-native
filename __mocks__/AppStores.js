export default {
  __esModule: true,
  getStores: jest.fn(),
};
