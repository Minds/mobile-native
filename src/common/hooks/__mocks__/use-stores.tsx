import { useLocalStore } from 'mobx-react';
import createPortraitStore from '../../../portrait/createPortraitStore';
import createSearchResultStore from '../../../topbar/searchbar/createSearchResultStore';
import createWalletStore from '../../../wallet/v2/createWalletStore';
import { createLegacyStores, TLegacyStores } from '../../contexts';
import { StoresType } from '../use-stores';

/**
 * Allows for a function component to consume our global stores
 * This **MUST** be only consumed below <StoresProvider> (which is placed in src/App.tsx)
 */
export const useStores = (): StoresType => {
  const stores = {
    wallet: useLocalStore(createWalletStore),
    searchBar: useLocalStore(createSearchResultStore),
    portrait: useLocalStore(createPortraitStore),
  };
  return stores;
};

/**
 * Allows for a function component to consume our legacy global stores that have been
 * provided via mobx's <Provider>
 */
export const useLegacyStores = (): TLegacyStores => {
  return createLegacyStores();
};
