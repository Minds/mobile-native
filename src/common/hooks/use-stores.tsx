import React from 'react';
import { TLegacyStores } from '../contexts';
import { useLocalStore, MobXProviderContext } from 'mobx-react';
import createWalletStore, {
  WalletStoreType,
} from '../../wallet/v2/createWalletStore';
import createSearchResultStore, {
  SearchResultStoreType,
} from '../../topbar/searchbar/createSearchResultStore';
import createPortraitStore, {
  PortraitStoreType,
} from '../../portrait/createPortraitStore';
import sessionService from '../services/session.service';
import logService from '../services/log.service';

export const storesContext = React.createContext<StoresType | null>(null);

/**
 * This is used in /src/App.tsx and provides a single instance of
 * our global stores
 */
export const StoresProvider = ({ children }) => {
  const stores = {
    wallet: useLocalStore(createWalletStore),
    searchBar: useLocalStore(createSearchResultStore),
    portrait: useLocalStore(createPortraitStore),
  };

  React.useEffect(() => {
    return sessionService.onLogout(() => {
      for (const id in stores) {
        if (stores[id].reset) {
          logService.info(`Reseting store ${id}`);
          stores[id].reset();
        }
      }
    });
  }, []);

  return (
    <storesContext.Provider value={stores}>{children}</storesContext.Provider>
  );
};

export type StoresType = {
  wallet: WalletStoreType;
  searchBar: SearchResultStoreType;
  portrait: PortraitStoreType;
};

/**
 * Allows for a function component to consume our global stores
 * This **MUST** be only consumed below <StoresProvider> (which is placed in src/App.tsx)
 */
export const useStores = (): StoresType => {
  const store = React.useContext(storesContext);
  if (!store) {
    // this is especially useful in TypeScript so you don't need to be checking for null all the time
    throw new Error('useStores must be used within a StoreProvider.');
  }
  return store;
};

/**
 * Allows for a function component to consume our legacy global stores that have been
 * provided via mobx's <Provider>
 */
export const useLegacyStores = (): TLegacyStores => {
  return React.useContext(MobXProviderContext) as TLegacyStores;
};
