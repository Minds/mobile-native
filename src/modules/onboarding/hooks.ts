import {
  ComponentOnboardingV5OnboardingStep,
  GetOnboardingV5VersionsQuery,
  useGetOnboardingV5VersionsQuery,
} from '~/graphql/strapi';

const carouselExtractor = (data?: GetOnboardingV5VersionsQuery) => {
  const screens = [
    ...new Map(
      data?.onboardingV5Versions?.data?.[0]?.attributes?.steps
        ?.map(step => (step as ComponentOnboardingV5OnboardingStep)?.carousel)
        .flat()
        .map(item => [item?.title, { ...item?.media, title: item?.title }]),
    ).values(),
  ];
  return screens;
};

type Survey = {
  title: string | undefined;
  description: string | undefined;
  radioSurveyQuestion: string | null | undefined;
  radioSurvey:
    | Array<{
        id: string;
        optionTitle: string;
        optionDescription: string;
        optionKey: string;
      } | null>
    | null
    | undefined;
};

const surveyExtractor = (
  data?: GetOnboardingV5VersionsQuery,
): Survey[] | undefined => {
  return data?.onboardingV5Versions?.data?.[0]?.attributes?.steps
    ?.filter(step => {
      const { radioSurvey, radioSurveyQuestion } =
        (step as ComponentOnboardingV5OnboardingStep) ?? {};
      return radioSurvey && radioSurveyQuestion;
    })
    .map(step => {
      const { title, description, radioSurvey, radioSurveyQuestion } =
        (step as ComponentOnboardingV5OnboardingStep) ?? {};
      return { title, description, radioSurvey, radioSurveyQuestion };
    });
};

export const useCarouselData = () => {
  const { data: originalData, ...rest } = useGetOnboardingV5VersionsQuery();
  const data = carouselExtractor(originalData);
  return {
    noData: data.length === 0,
    data,
    ...rest,
  };
};

export const useSurveyData = () => {
  const { data: originalData, ...rest } = useGetOnboardingV5VersionsQuery();
  const data = surveyExtractor(originalData);
  return {
    noData: (data?.length ?? 0) === 0,
    data,
    ...rest,
  };
};
