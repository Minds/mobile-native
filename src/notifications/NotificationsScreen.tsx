//@ts-nocheck
import React, { Component } from 'react';

import { Text, View, FlatList } from 'react-native';

import { observer, inject } from 'mobx-react';

import MIcon from 'react-native-vector-icons/MaterialIcons';

import { CommonStyle as CS } from '../styles/Common';
import { ComponentsStyle } from '../styles/Components';
import Notification from './notification/Notification';
import NotificationsTopbar from './NotificationsTopbar';
import ErrorBoundary from '../common/components/ErrorBoundary';
import i18n from '../common/services/i18n.service';
import Topbar from '../topbar/Topbar';
import ThemedStyles from '../styles/ThemedStyles';
import OnFocus from '../common/components/helper/OnFocus';
import CenteredLoading from '../common/components/CenteredLoading';

/**
 * Notification Screen
 */
@inject('notifications', 'user')
@observer
class NotificationsScreen extends Component {
  headerIndex = [0];

  /**
   * On screen focus
   */
  onFocus = () => {
    // when return to the screen reload only if there are new notifications
    if (this.props.notifications.unread > 0) {
      this.props.notifications.refresh();
      this.props.notifications.setUnread(0);
    }
  };

  /**
   * On component mount
   */
  componentDidMount() {
    this.disposeTabPress = this.props.navigation.addListener(
      //@ts-ignore
      'tabPress',
      this.refresh,
    );
  }

  /**
   * Initial load
   */
  async initialLoad() {
    try {
      await this.props.notifications.readLocal();
    } finally {
      await this.props.notifications.loadList(true);
    }
  }

  /**
   * On component unmount
   */
  componentWillUnmount() {
    // clear data to free memory
    this.props.notifications.list.clearList();

    if (this.disposeTabPress) {
      this.disposeTabPress();
    }
  }

  /**
   * Render screen
   */
  render() {
    const me = this.props.user.me;
    const list = this.props.notifications.list;
    let empty = null;

    if (list.loaded && !list.refreshing) {
      let design = null;

      if (me && me.hasBanner && !me.hasBanner()) {
        //TODO: check for avatar too
        design = (
          <Text
            style={ComponentsStyle.emptyComponentLink}
            onPress={() =>
              this.props.navigation.push('Channel', { username: 'me' })
            }>
            {i18n.t('newsfeed.designYourChannel')}
          </Text>
        );
      }

      empty = (
        <View style={ComponentsStyle.emptyComponentContainer}>
          <View style={ComponentsStyle.emptyComponent}>
            <MIcon name="notifications" size={72} color="#444" />
            <Text style={ComponentsStyle.emptyComponentMessage}>
              {i18n.t('notification.dontHave')}
            </Text>
            {design}
            <Text
              style={ComponentsStyle.emptyComponentLink}
              onPress={() => this.props.navigation.navigate('Capture')}>
              {i18n.t('createAPost')}
            </Text>
          </View>
        </View>
      );
    }

    return (
      <View style={CS.flexContainer}>
        <OnFocus onFocus={this.onFocus} />
        <FlatList
          data={list.entities.slice()}
          renderItem={this.renderRow}
          keyExtractor={this.keyExtractor}
          onRefresh={this.refresh}
          onEndReached={this.loadMore}
          ListEmptyComponent={this.props.notifications.loading ? null : empty}
          ListHeaderComponent={<NotificationsTopbar />}
          ListFooterComponent={
            this.props.notifications.loading && !list.refreshing ? (
              <CenteredLoading />
            ) : null
          }
          // onEndReachedThreshold={0.05}
          initialNumToRender={12}
          stickyHeaderIndices={this.headerIndex}
          windowSize={8}
          refreshing={list.refreshing}
          style={[ThemedStyles.style.backgroundPrimary, CS.flexContainer]}
        />
      </View>
    );
  }

  /**
   * Key extractor
   */
  keyExtractor = (item, index) =>
    `${item.time_created}:${item.from.guid}:${
      item.entity ? item.entity.guid + index : index
    }`;

  /**
   * Clear and reload
   */
  refresh = async () => {
    await this.props.notifications.loadRemoteOrLocal();
    this.props.notifications.setUnread(0);
  };

  /**
   * Load more rows
   */
  loadMore = () => {
    this.props.notifications.loadList();
  };

  /**
   * render row
   * @param {object} row
   */
  renderRow = (row) => {
    const entity = row.item;
    return (
      <ErrorBoundary
        message="Can't show this notification"
        containerStyle={CS.hairLineBottom}>
        <Notification entity={entity} navigation={this.props.navigation} />
      </ErrorBoundary>
    );
  };
}

export default NotificationsScreen;
