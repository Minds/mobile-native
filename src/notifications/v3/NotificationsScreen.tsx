import React from 'react';
import { observer } from 'mobx-react';
import { View, Text, FlatList, ViewToken } from 'react-native';
import ThemedStyles from '../../styles/ThemedStyles';
import NotificationsTopBar from './NotificationsTopBar';
import useApiFetch from '../../common/hooks/useApiFetch';
import i18n from '../../common/services/i18n.service';
import NotificationItem from './notification/Notification';
import { useStores } from '../../common/hooks/use-stores';
import ErrorBoundary from '../../common/components/ErrorBoundary';
import NotificationModel from './notification/NotificationModel';
import UserModel from '../../channel/UserModel';
import EmptyList from '../../common/components/EmptyList';
import NotificationPlaceHolder from './notification/NotificationPlaceHolder';

type PropsType = {
  navigation?: any;
};

const viewabilityConfig = {
  itemVisiblePercentThreshold: 50,
  minimumViewTime: 300,
  waitForInteraction: false,
};

type NotificationList = {
  status: string;
  notifications: NotificationModel[];
  'load-next': string;
};

const updateState = (newData: NotificationList, oldData: NotificationList) => {
  if (newData && newData.notifications) {
    newData.notifications = newData.notifications.map(
      (notification: NotificationModel) => {
        notification = NotificationModel.create(notification);
        if (notification.from) {
          notification.from = UserModel.create(notification.from);
        }
        if (notification.merged_from && notification.merged_from.length > 0) {
          notification.merged_from = UserModel.createMany(
            notification.merged_from,
          );
        }
        return notification;
      },
    );
    return {
      ...newData,
      notifications: [
        ...(oldData ? oldData.notifications : []),
        ...newData.notifications,
      ],
    } as NotificationList;
  }
};

const Empty = <EmptyList />;

const NotificationsScreen = observer(({ navigation }: PropsType) => {
  const theme = ThemedStyles.style;
  const { notifications } = useStores();
  const params = {
    filter: notifications.filter,
    limit: 15,
    offset: notifications.offset,
  };
  const {
    result,
    error,
    loading,
    fetch,
    setResult,
  } = useApiFetch<NotificationList>('api/v3/notifications/list', {
    params,
    updateState,
  });

  const onFetchMore = () => {
    !loading &&
      result &&
      result['load-next'] &&
      notifications.setOffset(result['load-next']);
  };

  const refresh = React.useCallback(() => {
    notifications.setOffset('');
    setResult(null);
    fetch(params);
  }, [notifications, setResult, fetch, params]);

  const onFocus = React.useCallback(() => {
    notifications.setUnread(0);
    refresh();
  }, [notifications, refresh]);

  //useFocusEffect(onFocus);

  React.useEffect(() => {
    const unsubscribe = navigation.addListener(
      //@ts-ignore
      'tabPress',
      onFocus,
    );

    return unsubscribe;
  }, [navigation, onFocus]);

  const headerComponent = React.useMemo(
    () => <NotificationsTopBar store={notifications} setResult={setResult} />,
    [notifications, setResult],
  );

  const onViewableItemsChanged = React.useCallback(
    (viewableItems: { viewableItems: ViewToken[]; changed: ViewToken[] }) => {
      viewableItems.viewableItems.forEach(
        (item: { item: NotificationModel }) => {
          if (!item.item.read) {
            item.item.read = true;
            notifications.markAsRead(item.item);
          }
        },
      );
    },
    [notifications],
  );

  const ListEmptyComponent = React.useMemo(() => {
    if (error && !loading) {
      return (
        <Text style={errorStyle} onPress={() => fetch()}>
          {i18n.t('cantReachServer') + '\n'}
          <Text style={[theme.colorLink, theme.marginTop2x]}>
            {i18n.t('tryAgain')}
          </Text>
        </Text>
      );
    }

    if (loading) {
      return (
        <View>
          <NotificationPlaceHolder />
          <NotificationPlaceHolder />
          <NotificationPlaceHolder />
          <NotificationPlaceHolder />
          <NotificationPlaceHolder />
          <NotificationPlaceHolder />
        </View>
      );
    } else {
      return Empty;
    }
  }, [error, loading, fetch]);

  const data = result?.notifications || [];

  return (
    <View style={theme.flexContainer}>
      <FlatList
        data={data.slice()}
        ListHeaderComponent={headerComponent}
        keyExtractor={keyExtractor}
        renderItem={renderItem}
        onEndReached={onFetchMore}
        onRefresh={refresh}
        refreshing={loading}
        onViewableItemsChanged={onViewableItemsChanged}
        viewabilityConfig={viewabilityConfig}
        ListEmptyComponent={ListEmptyComponent}
      />
    </View>
  );
});

const keyExtractor = (item: NotificationModel, index) => `${item.urn}-${index}`;

const renderItem = (row: any): React.ReactElement => {
  const notification = row.item;

  return (
    <ErrorBoundary
      message="Can't show this notification"
      containerStyle={ThemedStyles.style.borderBottomHair}>
      <NotificationItem notification={notification} />
    </ErrorBoundary>
  );
};

export default NotificationsScreen;

const errorStyle = ThemedStyles.combine(
  'colorSecondaryText',
  'textCenter',
  'fontXL',
  'marginVertical4x',
);
