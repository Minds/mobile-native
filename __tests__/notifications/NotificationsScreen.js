import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import NotificationsScreen from '../../src/notifications/v3/NotificationsScreen';
import { StoresProvider } from '../../src/common/hooks/use-stores';
import { getStores } from '../../AppStores';

jest.mock('@react-navigation/native');
jest.mock('react-native-system-setting');
jest.mock('react-native-notifications');
jest.mock(
  '../../src/common/components/interactions/InteractionsBottomSheet',
  () => 'InteractionsBottomSheet',
);

getStores.mockReturnValue({
  user: {
    me: {},
    load: jest.fn().mockReturnValue({ guid: '1' }),
    setUser: jest.fn(),
  },
});

describe('Notifications Screen Component', () => {
  let navigation;
  beforeEach(() => {
    navigation = {
      navigate: jest.fn(),
      push: jest.fn(),
      addListener: jest.fn(),
    };
  });
  it('renders correctly', () => {
    const { toJSON } = render(
      <StoresProvider>
        <NotificationsScreen navigation={navigation} />
      </StoresProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
