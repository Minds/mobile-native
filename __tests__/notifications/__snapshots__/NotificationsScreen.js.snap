// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Screen Component renders correctly 1`] = `
<View
  style={
    [
      {
        "alignContent": "center",
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    style={
      {
        "alignSelf": "stretch",
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#F2F2F2",
          },
          {
            "fontFamily": "Roboto_700Bold",
          },
          {
            "fontSize": 30,
            "lineHeight": 36,
          },
          {
            "textAlign": "left",
          },
        ]
      }
    >
      Sorry!
    </Text>
  </View>
  <View
    style={
      {
        "alignSelf": "stretch",
        "marginVertical": 12,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#F2F2F2",
          },
          {
            "fontFamily": "Roboto_700Bold",
          },
          {
            "fontSize": 18,
            "lineHeight": 28,
          },
          {
            "textAlign": "left",
          },
        ]
      }
    >
      Error displaying the content
    </Text>
  </View>
  <View
    style={
      {
        "alignSelf": "stretch",
        "marginBottom": 16,
      }
    }
  >
    <Text
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        {
          "bottom": 4,
          "left": 4,
          "right": 4,
          "top": 4,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "color": "#F2F2F2",
          },
          {
            "fontFamily": "Roboto_400Regular",
          },
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "textAlign": "center",
          },
        ]
      }
    >
      Tap to copy the error
      

      and
    </Text>
  </View>
  <View
    style={
      {
        "alignSelf": "stretch",
        "marginVertical": 16,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignSelf": "stretch",
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "flexDirection": "column",
              "justifyContent": "center",
              "overflow": "hidden",
            },
            undefined,
            {
              "minHeight": 42,
              "paddingHorizontal": 24,
              "paddingVertical": 2,
            },
            undefined,
            undefined,
            undefined,
            false,
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#fd8b7c",
              "borderColor": "transparent",
              "borderRadius": 100,
              "borderWidth": 1,
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(0, 0, 0, 0.1)",
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          collapsable={false}
          style={
            {
              "transform": [
                {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "color": "#FFFFFF",
                },
                {
                  "fontFamily": "Roboto_500Medium",
                },
                {
                  "fontSize": 18,
                  "lineHeight": 24,
                },
                {
                  "textAlign": "left",
                },
              ]
            }
          >
            Go back
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
